src/Device/Driver/AR62xx.cpp

```diff
@@ -0,0 +1,443 @@
+
+  return frequency_id;
+}
+
+/**
+ * @brief crc algorithm
+ * 
+ * @param data the data to calculate the crc on
+ * @param len the length of the data
+ * @return uint16_t the crc
+ */
+static uint16_t CRCBitwise(const uint8_t *data, const size_t len)
+{
+  uint16_t crc = 0x0000;
+  for (size_t j = len; j > 0; j--)
+  {
+    crc ^= (uint16_t)(*data++) << 8;
+    for (int i = 0; i < 8; i++)
+    {
+      if (crc & 0x8000)
+        crc = (crc << 1) ^ 0x8005;
+      else
+        crc <<= 1;
+    }
+  }
+  return crc;
+}
+
+unsigned int AR62xxDevice::GenerateSetStationCommand(uint8_t *command, unsigned int active_passive, const RadioFrequency frequency)
+{
+  /* length of the command */
+  unsigned int command_length = 0;
+
+  /* convert the active-frequency to an AR62xx-ID */
+  uint16_t active_frequency_idx = FrequencyToAR62xxId(active_frequency);
+
+  /* convert the passive-frequency to an AR62xx-ID */
+  uint16_t passive_frequency_idx = FrequencyToAR62xxId(passive_frequency);
+
+  /* command-start */
+  command[command_length++] = 0xA5;
+
+  /* add prot-ID */
+  command[command_length++] = 0x14;
+  command[command_length++] = 5;
+
+  /* set active station */
+  if (active_passive == ACTIVE_STATION)
+  {
+    active_frequency = frequency;
+    active_frequency_idx = FrequencyToAR62xxId(frequency);
+  }
+
+  /* set passive station */
+  else
+  {
+    passive_frequency = frequency;
+    passive_frequency_idx = FrequencyToAR62xxId(frequency);
+  }
+
+  /* setting frequencies */
+  command[command_length++] = 22;
+
+  //add active station
+  uint8_t tmp[2] = {};
+  WriteUnalignedLE16((uint16_t *)(void *)tmp, active_frequency_idx);
+  command[command_length++] = tmp[1];
+  command[command_length++] = tmp[0];
+
+  //add passive station
+  WriteUnalignedLE16((uint16_t *)(void *)tmp, passive_frequency_idx);
+  command[command_length++] = tmp[1];
+  command[command_length++] = tmp[0];
+
+  //add crc
+  WriteUnalignedLE16((uint16_t *)(void *)tmp, CRCBitwise(command, command_length));
+  command[command_length++] = tmp[1];
+  command[command_length++] = tmp[0];
+
+  //return full command
+  return command_length;
+}
+
+bool AR62xxDevice::ParseReceivedData(const uint8_t *data, size_t len)
+{
+  size_t idx = 0;
+  size_t rec_buf_len = 0;
+
+  /* we only need to read the message until record-size = 10 because then it already contains
+   * all data we want to read (active and passive frequency) */
+  size_t rec_bufsize = 10;
+  uint8_t command[rec_bufsize];
+
+  //flag if frequency has been saved
+  bool frequencySaved = false;
+
+  /* loop as long we have data and stop if frequency has been set */
+  while (!frequencySaved && rec_buf_len < rec_bufsize && idx < len)
+  {
+    /* check if we received a header which tells us now data starts */
+    if (data[idx] == 0xA5)
+    {
+      /* reset buffer length */
+      rec_buf_len = 0;
+    }
+
+    /* write command in buffer */
+    command[rec_buf_len++] = (char)data[idx++];
```

Comment:
Why this cast? And why copy data? What's the point of this stack buffer?

Reply:
solved... please have a look at the new implementation which is completely different from this and fully documented :-)